File: server/spring-boot-ecommerce/src/main/java/com/ecommerce/springbootecommerce/domain/Address.java
package com.ecommerce.springbootecommerce.domain;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.util.HashSet;
import java.util.Set;

@Getter
@Setter
@Entity
@Table(name="address")
public class Address {

    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE)
    private Long id;

    @ManyToOne
    @JoinColumn(name = "civility_id")
    private Civility civility;

    @Column(name="first_name")
    private String firstName;

    @Column(name = "last_name")
    private String lastName;

    private String street;

    @Column(name = "address_complement")
    private String addressComplement;

    private String city;

    @Column(name = "post_code")
    private int postCode;

    @ManyToOne
    @JoinColumn(name = "country_id")
    private Country country;

    @Column(name = "phone_number")
    private String phoneNumber;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;

    @OneToMany(mappedBy = "address", cascade = CascadeType.ALL)
    private Set<Order> orders = new HashSet<>();

}

File: client/ecommerce-app/src/app/models/address.ts
import { Country } from "./country";


interface AddressArgs {
  civility: number;
  firstName: string;
  lastName: string;
  street: string;
  addressComplement: string;
  postCode: number;
  city: string;
  country: Country;
  phoneNumber: string;
}

export class AddressBuilder {
  id!: number;
  civility: number;
  firstName: string;
  lastName: string;
  street: string;
  addressComplement: string;
  postCode: number;
  city: string;
  country: Country;
  phoneNumber: string;

  constructor(args: AddressArgs) {
    this.civility = args.civility;
    this.firstName = args.firstName;
    this.lastName = args.lastName;
    this.street = args.street;
    this.addressComplement = args.addressComplement;
    this.postCode = args.postCode;
    this.city = args.city;
    this.country = args.country;
    this.phoneNumber = args.phoneNumber;
  }
  
  private setId(id: number){
    this.id = id;
    return this;
  }

  buildNewAddress(){
    this.setId(0);
    return this.build();
  }

  buildExistingAddress(id: number){
    this.setId(id);
    return this.build();
  }

  build() {
    return new Address(
      this.id,
      this.civility,
      this.firstName,
      this.lastName,
      this.street,
      this.addressComplement,
      this.postCode,
      this.city,
      this.country,
      this.phoneNumber,
    );
  }
  
}

export class Address {
  id: number;
  civility: number;
  firstName: string;
  lastName: string;
  street: string;
  addressComplement: string;
  postCode: number;
  city: string;
  country: Country;
  phoneNumber: string;

  constructor(id: number, civility: number, firstName: string, lastName: string, street: string, addressComplement: string, postCode: number, city: string, country: Country, phoneNumber: string)
  {
    this.id = id;
    this.civility = civility;
    this.firstName = firstName;
    this.lastName = lastName;
    this.street = street;
    this.addressComplement = addressComplement;
    this.postCode = postCode;
    this.city = city;
    this.country = country;
    this.phoneNumber = phoneNumber;
  } 

}